### PR TITLE
igs017.cpp, igs022.cpp: Improved protection and decryption. Moved protection data to external files

### PR DESCRIPTION
**Pull request by LucaElia**
igs017.cpp: Improved protection and decryption [Luca Elia]
- Implemented "magic" i/o as a memory map (IGS_MUX) for added flexibility
- Hooked up IGS_INCDEC protection to cpoker2, sdmg2, tarzanc, spkrform, starzan
- Implemented specific IGS_INC protection in cpoker2
- Hooked up improved IGS022 protection to lhzb2, sqlz2
- Added IGS025 string protection to lhzb2, lhzb2a, mgcs, spkrform, slqz2, starzan, tarzanc, tjsb
- Palette scramble and tweaked tiles decryption in tarzanc (used also by starzan, happyskl, cpoker2)
- Decrypted sprites in tarzanc and starzan (used also by happyskl)
- Lamps and layout for starzan, happyskl, cpoker2
- Cleaned up/finished hopper emulation, added diplocations
- Joystick inputs in mgcs
- Finished inputs in spkrform. Allow hiding gambling (switching to Formosa and back)
- Moved protection data to external files
- Converted to logmacro

igs022.cpp: Fixes for igs017.cpp games [Luca Elia, RockyWall]
- Fixed initial auto-DMA mode
- Enlarged internal RAM. Added stack
- Fixed command 12: Copy -> Stack Push
- Added command 45: Stack Pop
- Fixed command 6d opcode 1: Add Imm -> Sub Values
- Added command 6d opcode 0: Add Values
- Extended logging and converted to logmacro

Machines promoted to working
----------------------------
Tarzan Chuang Tian Guan (China, V109C, set 1) [Luca Elia, iq_132, Ivan Vangelista, Guru, Dyq, bnathan]
Super Tarzan (Italy, V100I) [Luca Elia, iq_132, Ivan Vangelista, f205v, Mirko Buffoni]
Happy Skill (Italy, V611IT) [Luca Elia, Ivan Vangelista, Caius, The Dumping Union]
Champion Poker 2 (V100A) [Luca Elia, Ivan Vangelista, Jorge Silva, Fernando Oliveira]
Long Hu Zhengba 2 (China, set 1) [Luca Elia, RockyWall, David Haywood, iq_132, Pierpaolo Prazzoli, XingXing]
Shuang Long Qiang Zhu 2 VS (China, VS203J) [Luca Elia, RockyWall, David Haywood, iq_132, Pierpaolo Prazzoli, XingXing]

Clones promoted to working
--------------------------
Super Poker (V100xD03) / Formosa [Luca Elia, ANY]

### CODE REVIEW

**Review by cuavas — CHANGES_REQUESTED** [review]
This looks a lot better than last time.

There are a few `logerror` that are probably going to be a bit spammy and have a performance impact – think about which ones can be replaced with `LOG` from logmacro.h so they can be enabled conditionally.  It’s fine to leave in `logerror` for the ones for things that are not understood, but the ones for debugging the properly emulated functions probably don’t need to be active by default.

The other comments should all be pretty simple to address.  Thanks for sorting out the major issues.

File: src/mame/drivers/igs017.cpp [hunk 1]
```diff
@@ -52,18 +53,219 @@ To Do:
...
+
+	u16 m_string_word = 0, m_word = 0;
+	u8 m_string_offs = 0;
+};
+
+DEFINE_DEVICE_TYPE(IGS_STRING, igs_string_device, "igs_string", "IGS String Protection")
+
+igs_string_device::igs_string_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, IGS_STRING, tag, owner, clock)
+	, m_region_key(*this, "key")
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Changed to DEVICE_SELF

File: src/mame/drivers/igs017.cpp [hunk 2]
```diff
@@ -52,18 +53,219 @@ To Do:
...
+	m_word |= (BIT( x, 14) ^ BIT(m_string_word, 14)) << 15; // bit 15
+
+	// bit 0
+	const u16 bit0 = BIT(~x, 15) ^ BIT(x, 7);
+
+	// bit 0 is further xor'd with bit x of the data written to this address
+	const u16 xor0 = BIT(data, offset & 7);
+	m_word |= bit0 ^ xor0;
+
+	logerror("%s: exec bitswap on port %02x = %02x - bit0 %x, xor0 %x, word %04x -> %04x\n", machine().describe_context(), offset, data, bit0, xor0, x, m_word);
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Converted to logmacro.

File: src/mame/drivers/igs017.cpp [hunk 3]
```diff
@@ -52,18 +53,219 @@ To Do:
...
+	return;
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Yes, it slipped through. Removed.

File: src/mame/drivers/igs017.cpp [hunk 4]
```diff
@@ -52,18 +53,219 @@ To Do:
...
+
+	const u16 next_string_word  =   (m_string_word & (0xff << (8 - shift))) | (next_string_byte << shift);
+
+	logerror("%s: advance string offs %02x -> %02x, string data = %02x, string word %04x -> %04x\n",
+		machine().describe_context(), m_string_offs, next_string_offs, next_string_byte, m_string_word, next_string_word);
+
+	m_string_offs = next_string_offs;
+	m_string_word = next_string_word;
+
+	return 0x00;
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Changed to return space.unmap().

File: src/mame/drivers/igs017.cpp [hunk 5]
```diff
@@ -201,197 +369,90 @@ void igs_bitswap_device::device_start()
...
-	m_address = m_m3 = m_mf = 0;
+	m_m3 = m_mf = 0;
 	m_val = m_word = 0;
 }
 
-void igs_bitswap_device::address_w(u8 data)
+u8 igs_bitswap_device::result_r()
 {
-	m_address = data;
+	u8 res = bitswap<16>(m_val, 0,0,0,0,0,0,0,0, 5,2,9,7,10,13,12,15) & 0xff;
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Changed to bitswap<8>. I'm too used to the old BITSWAP16 macro I guess.

File: src/mame/drivers/igs017.cpp [hunk 6]
```diff
@@ -2510,63 +2872,63 @@ static INPUT_PORTS_START( iqblocka )
...
-	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_START2         ) PORT_NAME( "Play / Last Bet" ) // play current bet or, if null, bet as last time (2P C in test mode)
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_START2         ) PORT_NAME("Play / Last Bet") // play current bet or, if null, the last bet (2P C in test mode)
 	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_JOYSTICK_UP    ) PORT_PLAYER(2) // unused? shown in test mode
 	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_JOYSTICK_DOWN  ) PORT_PLAYER(2) // ""
 	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_JOYSTICK_LEFT  ) PORT_PLAYER(2) // ""
 	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_JOYSTICK_RIGHT ) PORT_PLAYER(2) // ""
-	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON4        ) PORT_NAME( "Hold 4 / Half Double" )        // (2P A in test mode)
-	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_BUTTON5        ) PORT_NAME( "Hold 5 / Tile in Double Up?" ) // (2P B in test mode)
+	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON4        ) PORT_NAME("Hold 4 / Half Double")        // (2P A in test mode)
+	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_BUTTON5        ) PORT_NAME("Hold 5 / Tile in Double Up?") // (2P B in test mode)
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Done.

File: src/mame/drivers/igs017.cpp [hunk 7]
```diff
@@ -2577,15 +2939,15 @@ static INPUT_PORTS_START( iqblocka )
...
-	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_BUTTON6        )                    // unused?      (1P E in test mode)
-	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_START3         ) PORT_NAME( "Bet" ) // Bet 1 credit (2P C in test mode)
-	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_BUTTON4        ) PORT_PLAYER(2)     // unused?      (2P D in test mode)
-	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON5        ) PORT_PLAYER(2)     // unused?      (2P E in test mode)
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_BUTTON3        ) PORT_NAME("Hold 3 / Small")
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_COIN2          ) PORT_IMPULSE(5)  // no coin. Hopper sensor? impulse prevents coin error in gambling mode (1P D in test mode)
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_BUTTON6        )                  // unused?      (1P E in test mode)
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_START3         ) PORT_NAME("Bet") // Bet 1 credit (2P C in test mode)
+	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_BUTTON4        ) PORT_PLAYER(2)   // unused?      (2P D in test mode)
+	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON5        ) PORT_PLAYER(2)   // unused?      (2P E in test mode)
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: Done.

File: src/mame/drivers/igs017.cpp [hunk 8]
```diff
@@ -3059,87 +3344,87 @@ INPUT_PORTS_END
...
-	PORT_DIPNAME( 0x80, 0x80, "Number Type" )
+	PORT_DIPNAME( 0x80, 0x80, "Number Type" ) PORT_DIPLOCATION("SW2:8")
 	PORT_DIPSETTING(    0x80, "Number" )
 	PORT_DIPSETTING(    0x00, "Tile" )
 
 	PORT_START("COINS")
-	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_CUSTOM   ) // hopper switch
-	PORT_BIT( 0x02, IP_ACTIVE_LOW,  IPT_SERVICE2  ) // shown in test mode
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_CUSTOM    ) PORT_READ_LINE_DEVICE_MEMBER("hopper", hopper_device, line_r) // hopper switch
+	PORT_BIT( 0x02, IP_ACTIVE_LOW,  IPT_SERVICE3  ) PORT_NAME("Clear") // shown in test mode
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/drivers/igs017.cpp]: There are actually the "Hide Gambling" feature. I've improved on it a bit and standardized to IPT_SERVICE1 to activate it, while also moving "Statistics" to IPT_GAMBLE_BOOK across games.

File: src/mame/machine/igs022.cpp [hunk 9]
```diff
@@ -101,166 +127,273 @@ void igs022_device::IGS022_do_dma(u16 src, u16 dst, u16 size, u16 mode)
...
+
+	if (m_stack_ptr > 0)
+		--m_stack_ptr;
+
+	return data;
+}
+
+std::string igs022_device::stack_as_string() const
+{
+	std::stringstream stream;
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/machine/igs022.cpp]: Correct. Changed to std::ostringstream.

File: src/mame/machine/igs022.cpp [hunk 10]
```diff
@@ -101,166 +127,273 @@ void igs022_device::IGS022_do_dma(u16 src, u16 dst, u16 size, u16 mode)
...
+	stream << "stack:";
+
+	for (int i = 0; i <= m_stack_ptr; ++i)
+		util::stream_format(stream, " %08x", m_stack[i]);
+
+	return stream.str();
```
[the remark this answers is not in the document]
Comment (LucaElia) [reply src/mame/machine/igs022.cpp]: Changed as suggested.

**Pull request merged**